File: 55-JumpGame/55-JumpGame.exs
defmodule Solution do
  @spec can_jump(nums :: [integer]) :: boolean
  def can_jump(nums), do: do_can_jump(nums, 0)

  defp do_can_jump([], _reach), do: true

  defp do_can_jump([n | rest], reach) do
    cond do
      reach < 0 ->
        false

      true ->
        new_reach = max(reach, n) - 1
        do_can_jump(rest, new_reach)
    end
  end
end

File: 55-JumpGame/55-JumpGame.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        max_idx = 0

        for i in range(n):
            if i <= max_idx:
                max_idx = max(max_idx, i + nums[i])
            else:
                return False
                
        return max_idx >= n - 1
